workflow/src/main/java/bio/terra/tanagra/workflow/BuildTextSearchInformation.java

```diff
@@ -0,0 +1,166 @@
+package bio.terra.tanagra.workflow;
+
+import com.google.api.services.bigquery.model.TableFieldSchema;
+import com.google.api.services.bigquery.model.TableRow;
+import com.google.api.services.bigquery.model.TableSchema;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryOptions;
+import org.apache.beam.sdk.options.Default;
+import org.apache.beam.sdk.options.Description;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptors;
+
+/**
+ * A batch Apache Beam pipeline for building a table that contains all text search information for
+ * an entity.
+ */
+public final class BuildTextSearchInformation {
```

Comment:
I'm anticipating that eventually we may want to stitch all of these together to create a combined indexing pipeline (i.e. where most of the parameters are common and/or the pulled dynamically from the underlay). I'm wondering whether there are any patterns we can establish now to make this easier later

Reply:
I agree about driving the indexing pipeline from the underlay. For indexing, the inputs we need right now are queries that return the:
- set of all entity instances (e.g. all conditions)
- set of all parent-child relationships between entity instances if there's a hierarchy
- set of all entity instance-text search strings if we want to support text search on the entity

For each query, we can either:
- generate it from the entity mapping in the underlay
- take a SQL query text file

I expect we'll need to support both methods. Specifying an entity mapping in the underlay seems cleaner and we can already do this in backend (e.g. generate a query for all conditions). But some things are too convoluted to specify in the underlay (e.g. parent-child relationships in the ingredient hierarchy exist across two separate tables), and I don't think we should force the underlay to express anything more than basic SQL syntax (i.e. no unions). Maybe one way to handle this is to allow users to specify a SQL query text file in the underlay?

RE patterns to establish now before we've coupled the indexing to the underlay: I think the main thing is to make sure the indexing input queries are general enough to apply to other non-OMOP datasets. So we don't want e.g. a workflow that takes a `domain_id` as input and slices the `concept` table by that domain. But a workflow that takes a parent-child query and flattens the hierarchy into ancestor-descendant or calculates the paths is okay.